x-pack/metricbeat/module/azure/billing/service.go

```diff
@@ -6,61 +6,95 @@ package billing
 
 import (
 	"context"
+	"fmt"
 	"time"
 
-	"github.com/Azure/go-autorest/autorest/date"
-
-	"github.com/Azure/azure-sdk-for-go/services/consumption/mgmt/2019-10-01/consumption"
-	//"github.com/Azure/azure-sdk-for-go/services/costmanagement/mgmt/2019-11-01/costmanagement"
-	"github.com/Azure/go-autorest/autorest/azure/auth"
-
 	"github.com/elastic/beats/v7/x-pack/metricbeat/module/azure"
 	"github.com/elastic/elastic-agent-libs/logp"
 
-	"github.com/Azure/azure-sdk-for-go/services/costmanagement/mgmt/2019-11-01/costmanagement"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/consumption/armconsumption"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/costmanagement/armcostmanagement"
 )
 
 // Service offers access to Azure Usage Details and Forecast data.
 type Service interface {
-	GetForecast(scope string, startTime, endTime time.Time) (costmanagement.QueryResult, error)
+	GetForecast(
+		scope string,
+		startTime,
+		endTime time.Time,
+	) (armcostmanagement.QueryResult, error)
 	GetUsageDetails(
 		scope string,
 		expand string,
 		filter string,
-		skipToken string,
-		top *int32,
-		metricType consumption.Metrictype,
+		metricType armconsumption.Metrictype,
 		startDate string,
-		endDate string) (consumption.UsageDetailsListResultPage, error)
+		endDate string,
+	) (armconsumption.UsageDetailsListResult, error)
 }
 
 // UsageService is a thin wrapper to the Usage Details API and the Forecast API from the Azure SDK for Go.
 type UsageService struct {
-	usageDetailsClient *consumption.UsageDetailsClient
-	forecastClient     *costmanagement.ForecastClient
+	usageDetailsClient *armconsumption.UsageDetailsClient
+	forecastClient     *armcostmanagement.ForecastClient
 	context            context.Context
 	log                *logp.Logger
 }
 
 // NewService builds a new UsageService using the given config.
 func NewService(config azure.Config) (*UsageService, error) {
-	clientConfig := auth.NewClientCredentialsConfig(config.ClientId, config.ClientSecret, config.TenantId)
-	clientConfig.AADEndpoint = config.ActiveDirectoryEndpoint
-	clientConfig.Resource = config.ResourceManagerEndpoint
-	authorizer, err := clientConfig.Authorizer()
+	cloudServicesConfig := cloud.AzurePublic.Services
+
+	if config.ResourceManagerEndpoint != "" && config.ResourceManagerEndpoint != azure.DefaultBaseURI {
+		cloudServicesConfig[cloud.ResourceManager] = cloud.ServiceConfiguration{
+			Audience: config.ResourceManagerEndpoint,
+			Endpoint: config.ResourceManagerEndpoint,
+		}
+	}
```

Comment:
I am not 100% we can use `config.ResourceManagerEndpoint` for both `Audience` and `Endpoint` as we were considering yesterday.

If the values provided by the SDK are the following:

https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azcore/arm/runtime/runtime.go#L11-L24 

we should consider adding a new optional configuration parameter for the audience.

Did you get the chance to verify what values the SDK uses at runtime?

Reply:
The values for the `cloud.ResourceManager` Azure Public service are `audience: https://management.core.azure.net`, `endpoint: https://management.azure.com`.

I've checked and the audience is basically the scope you want to use, for which the access token has permissions. The `core` in the URL shouldn't make any difference as I tested with `https://management.azure.com` as the audience and the results are the same.

Reply:
Does it work on the Azure public cloud using `https://management.azure.com` as value for both audience and endpoint?

Reply:
That's great.

The final step is deciding whether to expose the resource manager audience as a configuration option.

Keep current options:
- pro: configuration options are unchanged 
- con: edge cases may require a different audience 

Are there any downsides to exposing the resource manager audience as a configuration option?

Reply:
I don't see any downsides to exposing the audience as a package field in Kibana, maybe as a hidden input for users that need it.

Reply:
Yeah, we can offer a new `resource_manager_audience` option in the same way we offer `resource_manager_endpoint`. 

For example, `resource_manager_endpoint` is not even present in the default configuration for the [azure module](https://github.com/elastic/beats/blob/main/x-pack/metricbeat/modules.d/azure.yml.disabled), but it is  [documented](https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-module-azure.html#_module_specific_configuration_notes_2) and available for the people who need it.

Reply:
I am available for testing it when it's ready.